executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+            });
+    }
+
+    // Check that actions are valid (e.g., variables writen to only once, and only read after being written to)
+    fn is_valid(&self) -> bool {
+        let mut prev_writes = BTreeSet::new();
+        for action in &self.actions {
+            let writes = action.writes();
+            // writing to a variable already written?
+            if !writes.is_disjoint(&prev_writes) {
+                return false;
+            }
+            // reading a variable that was not written to?
+            if !action.reads().is_subset(&prev_writes) {
+                return false;
+            }
+            prev_writes.extend(writes);
+        }
+        true
+    }
+
+    // Execute the machine effects for the given the parameters
+    pub fn call<Q: QueryCallback<T>>(
+        &self,
+        mutable_state: &MutableState<'_, T, Q>,
+        params: &mut [LookupCell<T>],
+        mut data: CompactDataRef<'_, T>,
+    ) {
+        let row_offset = data.row_offset().try_into().unwrap();
+        let (data, known) = data.as_mut_slices();
+
+        let mut vars = vec![T::zero(); self.var_count];
+
+        let mut eval_stack = vec![];
+        for action in &self.actions {
+            match action {
+                InterpreterAction::AssignExpression(idx, e) => {
+                    let val = e.evaluate(&mut eval_stack, &vars[..]);
+                    vars[*idx] = val;
+                }
+                InterpreterAction::ReadCell(idx, c) => {
+                    vars[*idx] = data[index(
+                        self.first_column_id,
+                        self.column_count,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                    )];
+                }
+                InterpreterAction::ReadParam(idx, i) => {
+                    vars[*idx] = get_param(params, *i);
+                }
+                InterpreterAction::WriteCell(idx, c) => {
+                    set(
+                        self.first_column_id,
+                        self.column_count,
+                        data,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                        vars[*idx],
+                    );
+                }
+                InterpreterAction::WriteParam(idx, i) => {
+                    set_param(params, *i, vars[*idx]);
+                }
+                InterpreterAction::WriteKnown(c) => {
+                    set_known(
+                        self.first_column_id,
+                        self.column_count,
+                        known,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                    );
+                }
+                InterpreterAction::MachineCall(id, arguments) => {
+                    // we know it's safe to escape the references here, but the compiler doesn't, so we use unsafe
```

Comment:
What do you mean by "escape"? Is the unsafe part here that we need to know that the indices are pairwise disjoint?

Reply:
Especially after `https://github.com/powdr-labs/powdr/pull/2315`, each machine call has unique variables. This means you could store them directly in the arguments, which of course makes referencing them more complicated.

Another way to avoid the unsafe would be to mutably iterate over the variables, filter and map into LookupCells.

Reply:
The compiler doesn't like that we collect mut refs to the variables inside the map (e.g., would allow multiple mutable refs to the same var). There can be a _lot_ of variables, so I think I prefer the little bit of unsafe here

Reply:
Can you 1) explain a bit more why it needs unsafe and why it is safe and 2) add code in `is_valid` that checks that the indices in a machine call do not overlap?